### PR TITLE
Non fungible verses

### PR DESCRIPTION
**Pull request by crush-157**

### CODE REVIEW

**Comment by crush-157** [comment]
I wanted to get this into a (hospital) passable shape before I go on holiday.

File: native-image/non-fungible-verses/README.md [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+The aims of this lab are:
+1. To introduce you to some of the capabilities of the Oracle Cloud Infrastructure (OCI) always free tier.
+2. To show you how using [GraalVM Native Image](https://docs.oracle.com/en/graalvm/enterprise/22/docs/reference-manual/native-image/) can help you make the most of your free tier allocation.
+
+### OCI "Always Free" Tier
+
+[Oracle Cloud Infrastructure (OCI)](https://www.oracle.com/cloud/) is Oracle's next-generation cloud, designed to run any application, faster and more securely, for less.
+
+OCI has a free tier, which is composed of two parts:
+- A free trial of a broad set of OCI services limited by time (30 day) or consumption ($300) limit.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:19]:
```suggestion
- A free trial of a broad set of OCI services limited by time (30 days) or consumption ($300) limit.
```

File: native-image/non-fungible-verses/README.md [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+- A set of ["always free"](https://www.oracle.com/cloud/free/#always-free) services that you can use for an unlimited time.
+
+This lab is going to use two services from the "always free" tier to develop and deploy an application.
+
+We're going to
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:24]:
```suggestion
You're going to
```

File: native-image/non-fungible-verses/README.md [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+- develop and run the application on an ARM-based [Ampere A1 Compute](https://www.oracle.com/cloud/compute/arm/ VM.
+- use the OCI [Autonomous JSON Database](https://www.oracle.com/autonomous-database/autonomous-json-database/) service to store the application's data.
+
+The application is built using Spring Boot.  It procedually generates "poems", and stores them as JSON documents in the autonomous JSON database (AJD).  We can make each generated poems publicly accessible via a unique URL.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:28]:
```suggestion
The application is built using Spring Boot.  It procedually generates "poems", and stores them as JSON documents in the autonomous JSON database (AJD).  Each generated poem is publicly accessible via a unique URL.
```

File: native-image/non-fungible-verses/README.md [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+
+For a full Web3 experince you can let people pay for the privilege of recording a transaction on a blockchain that includes the URL of a "poem", thereby "owning" the "digital collectible" non - fungible verse (NFV).
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:30]:
```suggestion
For a full Web3 experince you can let people pay for the privilege of recording a transaction on a blockchain that includes the URL of a "poem", thereby "owning" the "digital collectible" non-fungible verse (NFV).
```

File: native-image/non-fungible-verses/README.md [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+
+Create a compute instance
+- shape: VM.Standard.A1.Flex
+- OCPU: 2
+- RAM: 12GB
+- Image: vanilla Oracle Linux 8
+
+Once the VM is up and running, SSH into it and carry out the following steps:
+* Install [GraalVM version 22 or greater, JDK17+](https://www.graalvm.org/downloads/) - You can use either the Community or Enterprise Edition 
+* Install the `native-image` tool (see [Native Images](https://www.graalvm.org/22.0/docs/getting-started/#native-images))
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:100]:
```suggestion
* Install the `native-image` tool (see [Native Image](https://www.graalvm.org/22.0/docs/getting-started/#native-image))
```

File: native-image/non-fungible-verses/README.md [hunk 6]
```diff
@@ -0,0 +1,785 @@
...
+* In your `.bashrc` file, set `JAVA_HOME` to point to your GraalVM installation
+* stop & disable the firewall of the VM (normally this isn't recommended)
+* In the security list for the VCN Subnet of your VM, add an ingress rule for TCP traffic on port 8080.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:103]:
```suggestion
* In the security list for the VCN Subnet of your VM, add an ingress rule for TCP traffic on port 8080
```

File: native-image/non-fungible-verses/README.md [hunk 7]
```diff
@@ -0,0 +1,785 @@
...
+
+### Autonomous JSON Database
+
+Create an AJD instance with the default settings.
+
+Note your admin password.
+
+Once the database is up and running, click on "Database Actions".
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:111]:
```suggestion
Once the database is up and running, click *Database Actions*.
```

File: native-image/non-fungible-verses/README.md [hunk 8]
```diff
@@ -0,0 +1,785 @@
...
+
+In the "Database Actions" tab, click on "Restful Services and SODA".  Copy the base URL.  In the `.bashrc` file of the compute VM, set `$ORDS_BASE_URL` to the base URL.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:113]:
```suggestion
In the *Database Actions* tab, click *Restful Services and SODA*.  Copy the base URL.  In the `.bashrc` file of the compute VM, set `$ORDS_BASE_URL` to the base URL.
```

File: native-image/non-fungible-verses/README.md [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+
+Click on "Database Users".  Create a new user.  In the `.bashrc` file of the compute VM, set `$ORDS_USER` and `$ORDS_PASSWORD` to the username and password of the user you have created.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:115]:
```suggestion
Click *Database Users*.  Create a new user.  In the `.bashrc` file of the compute VM, set `$ORDS_USER` and `$ORDS_PASSWORD` to the username and password of the user you have created.
```

File: native-image/non-fungible-verses/README.md [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+
+Log out of "Database Actions", and back in as the user you just created.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:117]:
```suggestion
Log out of *Database Actions*, and then log in as the user you just created.
```

File: native-image/non-fungible-verses/README.md [hunk 11]
```diff
@@ -0,0 +1,785 @@
...
+
+Click on JSON.  Create a collection.  In the `.bashrc` file of the compute VM, set `$JSON_COLLECTION_NAME` to the name of the collection you just created.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:119]:
```suggestion
Click *JSON*.  Create a collection.  In the `.bashrc` file of the compute VM, set `$JSON_COLLECTION_NAME` to the name of the collection you just created.
```

File: native-image/non-fungible-verses/README.md [hunk 12]
```diff
@@ -0,0 +1,785 @@
...
+Create a compute instance
+- shape: VM.Standard.A1.Flex
+- OCPU: 2
+- RAM: 12GB
+- Image: vanilla Oracle Linux 8
+
+Once the VM is up and running, SSH into it and carry out the following steps:
+* Install [GraalVM version 22 or greater, JDK17+](https://www.graalvm.org/downloads/) - You can use either the Community or Enterprise Edition 
+* Install the `native-image` tool (see [Native Images](https://www.graalvm.org/22.0/docs/getting-started/#native-images))
+* In your `.bashrc` file, set `JAVA_HOME` to point to your GraalVM installation
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:101]:
```suggestion
* In your _.bashrc_ file, set `JAVA_HOME` to point to your GraalVM installation
```

File: native-image/non-fungible-verses/README.md [hunk 13]
```diff
@@ -0,0 +1,785 @@
...
+
+In the "Database Actions" tab, click on "Restful Services and SODA".  Copy the base URL.  In the `.bashrc` file of the compute VM, set `$ORDS_BASE_URL` to the base URL.
+
+Click on "Database Users".  Create a new user.  In the `.bashrc` file of the compute VM, set `$ORDS_USER` and `$ORDS_PASSWORD` to the username and password of the user you have created.
+
+Log out of "Database Actions", and back in as the user you just created.
+
+Click on JSON.  Create a collection.  In the `.bashrc` file of the compute VM, set `$JSON_COLLECTION_NAME` to the name of the collection you just created.
+
+The `.bashrc` file of the compute VM should look something like:
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:121]:
```suggestion
The contents of the _.bashrc_ file of the compute VM should look something like:
```

File: native-image/non-fungible-verses/README.md [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+export ORDS_USER="cyberpoet"
+export ORDS_PASSWORD="September_2022"
+export JSON_COLLECTION_NAME="poetry"
+```
+
+In your SSH terminal(s) run `source .bashrc`.
+
+## **STEP 1**: Meet Our Sample Java Application
+
+In this lab you are going to build a simple application with a very minimal REST-based API. You are then going to
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:154]:
```suggestion
This lab describes how to build a simple application with a very minimal REST-based API.  
```

File: native-image/non-fungible-verses/README.md [hunk 15]
```diff
@@ -0,0 +1,785 @@
...
+export ORDS_PASSWORD="September_2022"
+export JSON_COLLECTION_NAME="poetry"
+```
+
+In your SSH terminal(s) run `source .bashrc`.
+
+## **STEP 1**: Meet Our Sample Java Application
+
+In this lab you are going to build a simple application with a very minimal REST-based API. You are then going to 
+containerise the application, using Docker. First, take a quick look at your simple application.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:155]:
```suggestion
The lab then shows how to containerise the application, using Docker. First, take a quick look at your simple application.
```

File: native-image/non-fungible-verses/README.md [hunk 16]
```diff
@@ -0,0 +1,785 @@
...
+
+We have provided the source code and build scripts for this application in _native-image/containerisation/lab/src_.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:157]:
```suggestion
The source code and build scripts for this application are in _native-image/containerisation/lab/src_.
```

File: native-image/non-fungible-verses/README.md [hunk 17]
```diff
@@ -0,0 +1,785 @@
...
+by using a [Markov Chain](https://en.wikipedia.org/wiki/Markov_chain) to produce a model of the original poem (this is essentially a statistical model). 
+This model is then used to generate new text.
+
+If you call the endpoint `/mint` it will write a new "poem" to the collection in the AJD.
+
+If you call the endpoint `/read` it will retrieve all the poems from the collection.
+
+If you call the endpoint `/read/{uuid}` it will retrieve the poem with the specified `uuid` if it exists.
+
+In the example application, you provide it with the text of the poem, then generate a model of the text which the application then uses to
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:182]:
```suggestion
In the example application, you provide it with the text of the poem, then it generates a model of the text which the application then uses to 
```

File: native-image/non-fungible-verses/README.md [hunk 18]
```diff
@@ -0,0 +1,785 @@
...
+This model is then used to generate new text.
+
+If you call the endpoint `/mint` it will write a new "poem" to the collection in the AJD.
+
+If you call the endpoint `/read` it will retrieve all the poems from the collection.
+
+If you call the endpoint `/read/{uuid}` it will retrieve the poem with the specified `uuid` if it exists.
+
+In the example application, you provide it with the text of the poem, then generate a model of the text which the application then uses to 
+generate a new text that is similar to the original text. We are using the [RiTa](https://rednoise.org/rita/) library to do the heavy lifting for us--it supports building and using Markov Chains.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:183]:
```suggestion
generate a new text that is similar to the original text. The [RiTa](https://rednoise.org/rita/) library performs the heavy lifting for us--it supports building and using Markov Chains.
```

File: native-image/non-fungible-verses/README.md [hunk 19]
```diff
@@ -0,0 +1,785 @@
...
+
+Below are two snippets from the utility class `com.example.demo.Jabberwocky` that builds the model. The `text` variable
+contains the text of the original poem. This snippet shows how we create the model and then populate it with `text`.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:186]:
```suggestion
contains the text of the original poem. This snippet shows how to create the model and then populate it with `text`.
```

File: native-image/non-fungible-verses/README.md [hunk 20]
```diff
@@ -0,0 +1,785 @@
...
+This is called from the class constructor
+and we define the class to be a [Singleton](https://docs.spring.io/spring-framework/docs/3.0.0.M3/reference/html/ch04s04.html#beans-factory-scopes-singleton)
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:188]:
```suggestion
and the class is defined to be a [Singleton](https://docs.spring.io/spring-framework/docs/3.0.0.M3/reference/html/ch04s04.html#beans-factory-scopes-singleton)
```

File: native-image/non-fungible-verses/README.md [hunk 21]
```diff
@@ -0,0 +1,785 @@
...
+# Bring the application back to the foreground
+fg
+# Terminate it with : <ctrl-c>
+<ctrl-c>
+```
+
+## **STEP 2**: Containerising Your Java Application with Docker
+
+Containerising your Java application as a Docker container is straightforward. You can build
+a new Docker image based on one that contains a JDK distribution. So, for this tutorial you will use a container
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:277]:
```suggestion
a new Docker image based on one that contains a JDK distribution. So, for this lab you will use a container
```

File: native-image/non-fungible-verses/README.md [hunk 22]
```diff
@@ -0,0 +1,785 @@
...
+```
+
+You should see your newly built image listed. Run this image as follows:
+
+![](images/RMIL_Technology_Laptop_Bark_RGB_50.png#input)
+```bash
+docker run --rm -d --name "jibber-jdk" -p 8080:8080 jibber:jdk.01
+```
+
+Then call the endpoint as you did before--you may need to wait for a few seconds before doing to allow the
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:328]:
```suggestion
Then call the endpoint as you did before--you may need to wait for a few seconds before doing so to allow the
```

File: native-image/non-fungible-verses/README.md [hunk 23]
```diff
@@ -0,0 +1,785 @@
...
+
+## **STEP 3**: Building a Native Executable
+
+So far you have:
+
+1. Built a Spring Boot application with a HTTP endpoint, `/jibber`
+2. Successfully containerised it
+
+Now you can create a native executable from your application using GraalVM Native Image. The native executable
+is going to two interesting characteristics, namely:
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:378]:
```suggestion
is going to exhibit two interesting characteristics, namely:
```

File: native-image/non-fungible-verses/README.md [hunk 24]
```diff
@@ -0,0 +1,785 @@
...
+echo $((size_in_bytes/1024/1024))
+```
+
+In our experiments, we saw a size of around 82MB, compared to 171MB for an Oracle Linux 8 
+image. So we have shrunk the container by a lot. A long way down from our starting size, for 
+the Java container, of almost 600MB.
+
+## Summary
+
+We hope you have enjoyed this lab and learnt a few things along the way. We've looked at how you can containerise
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:774]:
```suggestion
We hope you have enjoyed this lab and learnt a few things along the way. You've seen how to containerise
```

File: native-image/non-fungible-verses/README.md [hunk 25]
```diff
@@ -0,0 +1,785 @@
...
+a Java application. Then we've seen how to convert that Java application into a native executable, which starts significantly
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:775]:
```suggestion
a Java application. Then you've converted that Java application into a native executable, which starts significantly 
```

File: native-image/non-fungible-verses/README.md [hunk 26]
```diff
@@ -0,0 +1,785 @@
...
+the Java container, of almost 600MB.
+
+## Summary
+
+We hope you have enjoyed this lab and learnt a few things along the way. We've looked at how you can containerise
+a Java application. Then we've seen how to convert that Java application into a native executable, which starts significantly 
+faster than the Java application. You then containerised the native executable and have seen how the size of the 
+Docker image, with your native executable in it, is much smaller than the Java Docker Image.
+
+Finally, we looked at how we can build a mostly-statically linked native executable with Native Image. These can be
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:779]:
```suggestion
Finally, you saw how to build a mostly-statically linked native executable with Native Image. These can be
```

File: native-image/non-fungible-verses/README.md [hunk 27]
```diff
@@ -0,0 +1,785 @@
...
+packaged in smaller containers, such as Distroless and these let us shrink the size of the Docker Image even further.
```
Comment (bhoran) [inline native-image/non-fungible-verses/README.md:780]:
```suggestion
packaged in smaller containers, such as Distroless and these let you shrink the size of the Docker Image even further.
```

**Pull request merged**